File: solutions/java/src/main/java/com/epi/RebuildBSTPostorder.java
package com.epi;

import com.epi.BinaryTreePrototypeTemplate.BinaryTree;

import java.util.ArrayList;
import java.util.List;

/**
 * @author translated from c++ by Blazheev Alexander
 */
public class RebuildBSTPostorder {
  // @include
  // Given a postorder traversal of a BST, return its root.
  public static <T extends Comparable<T>> BinaryTree<T> rebuildBSTFromPostorder(
      List<T> postorder) {
    return rebuildBSTPostorderHelper(postorder, 0, postorder.size());
  }

  // Build a BST based on postorder[s : e - 1], return its root.
  private static <T extends Comparable<T>> BinaryTree<T> rebuildBSTPostorderHelper(
      List<T> postorder, int s, int e) {
    if (s < e) {
      int x = s;
      while (x < e && postorder.get(x).compareTo(postorder.get(e - 1)) < 0) {
        ++x;
      }
      return new BinaryTree<>(postorder.get(e - 1), rebuildBSTPostorderHelper(
          postorder, s, x), rebuildBSTPostorderHelper(postorder, x, e - 1));
    }
    return null;
  }
  // @exclude

  private static <T extends Comparable<T>> void checkAns(BinaryTree<T> n, T pre) {
    if (n != null) {
      checkAns(n.getLeft(), pre);
      assert (pre.compareTo(n.getData()) <= 0);
      System.out.println(n.getData());
      checkAns(n.getRight(), n.getData());
    }
  }

  public static void main(String[] args) {
    // 1
    // 2
    // 3
    // 4
    // 5
    // 6
    // should output 1, 2, 3, 4, 5, 6
    // postorder [6, 5, 4, 3, 2, 1]
    List<Integer> postorder = new ArrayList<>();
    postorder.add(6);
    postorder.add(5);
    postorder.add(4);
    postorder.add(3);
    postorder.add(2);
    postorder.add(1);
    BinaryTree<Integer> root = rebuildBSTFromPostorder(postorder);
    checkAns(root, Integer.MIN_VALUE);
  }
}

File: static/solutions/cpp/Rebuild_BST_postorder.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <limits>
#include <memory>
#include <string>
#include <vector>

#include "./BST_prototype.h"

using std::cout;
using std::endl;
using std::make_unique;
using std::numeric_limits;
using std::unique_ptr;
using std::vector;

unique_ptr<BSTNode<int>> RebuildBSTFromPostorderHelper(const vector<int>&,
                                                       int, int);

// @include
// Given a postorder traversal of a BST, return its root.
unique_ptr<BSTNode<int>> RebuildBSTFromPostorder(
    const vector<int>& postorder) {
  return RebuildBSTFromPostorderHelper(postorder, 0, postorder.size());
}

// Build a BST based on postorder[start : end - 1], return its root.
unique_ptr<BSTNode<int>> RebuildBSTFromPostorderHelper(
    const vector<int>& postorder, int start, int end) {
  if (start < end) {
    int x = start;
    while (x < end && postorder[x] < postorder[end - 1]) {
      ++x;
    }
    return make_unique<BSTNode<int>>(
        BSTNode<int>{postorder[end - 1],
                     RebuildBSTFromPostorderHelper(postorder, start, x),
                     RebuildBSTFromPostorderHelper(postorder, x, end - 1)});
  }
  return nullptr;
}
// @exclude

template <typename T>
void CheckAns(const unique_ptr<BSTNode<T>>& n, const T& pre) {
  if (n) {
    CheckAns(n->left, pre);
    assert(pre <= n->data);
    cout << n->data << endl;
    CheckAns(n->right, n->data);
  }
}

int main(int argc, char* argv[]) {
  //      3
  //    2   5
  //  1    4  6
  // should output 1, 2, 3, 4, 5, 6
  // preorder [1, 2, 4, 6, 5, 3]
  vector<int> postorder = {1, 2, 4, 6, 5, 3};
  unique_ptr<BSTNode<int>> root(RebuildBSTFromPostorder(postorder));
  CheckAns(root, numeric_limits<int>::min());
  return 0;
}
